### ISSUE TITLE
Playback of 24bit FLAC is not bit-perfect despite 16bit FLAC being bit-perfect

### ISSUE THREAD

**Issue by NintendoManiac64** [renpy/renpy#3050]
**2026 UPDATE**: As of Ren'Py 8.2 (see #6488 - it's not mentioned in changelogs), even 16bit FLAC can no longer be played back bit-perfect due to a hard-coded ~3dB volume reduction. ~~And although it's "perceptually lossless" due to Ren'Py now thankfully using a 32bit floating point audio mixer, the use of floating point does unfortunately mean that it's basically impossible for the end-user to achieve bit-perfect output since all you have access to is what amounts to a separate post-reduction volume amplifier.~~ **EDIT:** Subsequent recording tests show Ren'Py seemingly still internally reading and/or rendering audio at 16bit. Perhaps only gain adjustment is handled with 32bit precision?

## Original post

I know I know, this is some super-duper OCD audiophile crap that nobody in their right mind should ever care about, but I'm concerned about the possibility of a bug in how Ren'Py handles greater-than-16bit audio (which is actually pretty common with lossy formats since the likes of MP3, Vorbis, and Opus store their audio data as floating point).

Basically, while doing some audio tests to see how Ren'Py handles different audio formats, I discovered that it supports both 16bit and 24bit FLAC

...yet, when actually doing an audio recording to compare how Ren'Py renders the audio compared to the source files, only the recording for the 16bit FLAC ended up being 100% identical to the according FLAC file while the recording for 24bit FLAC ended up being a teeny bit different - almost like it was being reduced in bit depth to something between 24bit and 16bit, like 20bit or something.

Again, I realize this is super-duper unimportant, but the inconsistency in audio rendering makes me wonder if there's a bug or something.  I mean, if Ren'Py doesn't handle 24bit audio output then I wouldn't expect it to even support 24bit FLAC files, and yet it does.  This could arguably make sense if it just always truncated or down-sampled everything to 16bit audio output, but doesn't explain why the resulting audio render seemed to still be so different from the 16bit equivalent.

　

For reference I took a 192kHz 24bit copy of Simon & Garfunkel's "The Boxer" and downsampled it using foobar2000 and the SoX plug to both 48kHz 24bit and 48kHz 16bit.  For recording I just used Audacity (though admittedly an older version from before the UI change - v2.1.2) set to record "speakers" on my Xonar DS sound card which was similarly set to 48kHz 24bit.  All volume sliders in Windows, Audacity, and Ren'Py were set to 100%.

To compare the recording to the according FLAC I lined up the waveforms to the exact identical sample, inverted one of the waveforms, mixed the two together to a new waveform, and then looked at the value listed in "amplify" to see the difference - the larger the value, the smaller the difference.

Comparing the 16bit FLAC to its according recording returned an amplify value of "Infinity" (identical) while comparing the 24bit FLAC to its according recording resulted in an amplify value of 50dB with "New peak" listed as -40dB (so basically 90dB).  Comparing the 16bit FLAC to the recording of the 24bit FLAC returned an amplify value of 50dB with "New peak" listed as -13dB (so basically 63dB).

**Comment by renpytom** [comment]
Ren'Py only support 16 bit audio playback, and should downsample.

**Comment by NintendoManiac64** [comment]
> Ren'Py only support 16 bit audio playback, and should downsample.

Do you know if the internal volume slider also only operates at 16bit?

I ask because the audio mixer/renderer, including the volume slider, in Windows since Vista has been 32bit float and only dithers down to 24bit or 16bit for the final output.

**Comment by uyjulian** [comment]
Ren'Py is hardcoded to 16 bits output and decode. https://github.com/renpy/renpy/blob/master/module/ffmedia.c#L730  
https://github.com/renpy/renpy/blob/49e4a115e8a2e32372817047b041b7c2884ef385/module/renpysound_core.c#L1162

**Comment by uyjulian** [comment]
> I'm also guessing that simply changing those two specific "16" numerical values in the Ren'Py code to something like "24" would cause horrible breakage elsewhere in the codebase?

You would probably need to change buffer sizes too...

https://github.com/renpy/renpy/blob/2fbb7be5be6069a1dca5b10db73e20fd06495af8/module/ffmedia.c#L26  
https://github.com/renpy/renpy/blob/2fbb7be5be6069a1dca5b10db73e20fd06495af8/module/ffmedia.c#L29  
https://github.com/renpy/renpy/blob/2fbb7be5be6069a1dca5b10db73e20fd06495af8/module/ffmedia.c#L30  

However, this may not be all the locations that need to be changed.  I haven't actually tested this myself.

**Comment by renpytom** [comment]
Is there some documentation about mobile OSes supporting 32-bit? I'm not going to to move Ren'Py to a  floating point audio format, but s16 integer seems reasonable. Also, note that this is meant for a game engine, so I'm not going to judge audio playback by the standards of audiophile applications.

**Comment by NintendoManiac64** [comment]
So I was slightly off - iOS uses 24bit integer and 16bit integer, but possibly for different things? My major lack of experience with Apple stuff is making me uncertain as to what they're actually referring to:
https://developer.apple.com/library/archive/documentation/MusicAudio/Conceptual/CoreAudioOverview/CoreAudioEssentials/CoreAudioEssentials.html#//apple_ref/doc/uid/TP40003577-CH10-SW16

But Android by comparison is straight-up 32float nowadays:
https://source.android.com/devices/audio/highres-effects


So I would imagine that whether iOS would allow 24bit or if it requires 16bit from apps is the bottleneck here.


And yes 16bit integer is "reasonable" but it's kind like having our monitor set to a 60Hz refresh rate back in the days of CRTs rather than 75Hz despite AFAIK basically all CRT monitors also supporting 75Hz (which of course on a CRT also reduced the flicker) - outside of actually needing to actually put in a bit of time to change it, there's not really much (if any?) downside except _maybe_ over so slightly increased CPU time? (at least for floating point; I'm not even sure if 24bit integer would use more CPU than 16bit integer when all of these CPUs are at least 32bit architectures anyway unless the compiler was specifically able to fit two 16bit integer operations within a single 32bit integer operation).

**Comment by NintendoManiac64** [comment]
According to issue #6488 (and not mentioned anywhere in the Ren'Py changelogs...), audio mixing is all done in 32bit float as of Ren'Py v8.2 from 2024.

　

However, also according to it issue #6488 (and similarly not mentioned in the changelogs), it additionally seems that Ren'Py always applies what my tests show to be a 3dB reduction in volume, making it rather difficult to test for exact bit-perfect audio output.

Now I realize that, once you're doing everything in 32bit floating point, getting bit-perfect audio output really is kind of of no benefit. In fact, when I did recording tests with the same "mix the recording with an inverted source" technique, the floating point output was clearly being recorded as I was able to amplify audio that was -40dB *after* said mix&invert technique and have a result that sounded pretty much identical to the source - something you simply can't get with 16bit integer audio output!

(amusingly though, Ren'Py still can't handle 24bit and 32bit LPCM WAV files :P So I'm still using 24bit FLAC for all of my tests, though FLAC finally just in the last year or so also gained encoder support for 32bit integer so that's an interesting thing I may experiment with at some point, but because 32bit float has 24 bits of integer it'd be more a test of Ren'Py's audio dithering capabilities)

　

However, I cannot help but notice that it feels like the process as of Ren'py v8.4.1 (which introduced the `adjust_audio_amplitude` located in the user-accessable `renpy/config.py` file) is doing a case of "hard-code a volume reduction and then add a second user-facing volume control" which feels a bit redundant and "reinventing the wheel" all things considered.

Therefore, would it not just be simpler to ditch the hard-coded volume reduction altogether and simply have `adjust_audio_amplitude` set to a value of 0.7 by default and rely on that? Then all it would take is to set it to 1.0 and Bob's your uncle to bypass any volume processing that Ren'Py is doing "secretly" behind-the-scenes, and by definition presumably also allowing for bit-perfect playback if desired (when a given Ren'Py application's volume slider is set to 100% of course).

And even though 32bit floating point audio mixing effectively makes gain adjustment "perceptually lossless", it still can be useful for debugging and "sanity-check" purposes to at least have some method that bypasses all post-processing for a mathematically lossless output rather than just a perceptual one.

**Comment by renpytom** [comment]
Ren'Py is a game engine, and so bit-perfect audio is not a goal. The volume drop is there to support panning with no power loss, and is part of the panning mechanism, so it's not going to  be removed. 

I'm going to close this issue now, as I think it's been satisfied.

**Comment by NintendoManiac64** [comment]
> The volume drop is there to support panning with no power loss, and is part of the panning mechanism, so it's not going to be removed.

Maybe I didn't convey myself well enough. I'm now not so much asking to undo the v8.2 volume reduction, but rather it looks to me like Ren'Py currently has redundant audio processes that achieve said result? And maybe I have no idea what I'm talking about, but my perception thus far is that `adjust_audio_amplitude` does everything and more that the original v8.2 method of volume reduction achieves.

To be clear, I understand *why* the volume reduction is there, but I'm confused as to why there needs to what amounts to two separate volume controls in the back-end. I realize the v8.2 volume reduction was implemented x.2 versions before the `adjust_audio_amplitude` was implemented, but doesn't the subsequent addition of `adjust_audio_amplitude` completely negate any reason to still retain the original v8.2 method of volume reduction by virtue of redundancy?

**tl;dr** - what's the benefit of retaining *both* the original v8.2 method of volume reduction and `adjust_audio_amplitude`?

(extra clarification: whenever I'm mentioning `adjust_audio_amplitude` I am basing it on the idea that its default value is something less than 1.0 in order to facilitate the aforementioned volume reduction required to support panning with no power loss)

**Comment by NintendoManiac64** [comment]
Um, I don't want to be the person saying "you can't close this issue - it's important to me, so it MUST be important to you too!" but I think this may in fact still be an issue in that I think audio is still being internally read and/or rendered at 16 bit or something. Perhaps only gain adjustment is handled with 32bit floating point precision?

　

What I did is, once again, I took Simon & Garfunkel's "The Boxer" (which is my go-to native 192kHz 24bit FLAC test file), set Ren'Py's output sample rate to 192000 Hz and simply made an audio recording. Then I did it again but this time with a 16bit FLAC. From there I took the two recordings and, after normalizing the gain, I compared each to the source 24bit FLAC via the usual "align the samples exactly, inverting one waveform, mixing them together, and note the resulting "new peak amplitude" value in `Amplify`":

24bit: -26.745dB
16bit: -23.339dB

But to sanity-check what a proper result *should* be, I recreated in Audacity what Ren'Py is supposedly doing with it's 3dB reduction whereby I imported the source 24bit FLAC, reduced the gain by 3dB, and exported as 32bit, 24bit, and 16bit. Then after normalizing gain across the board, I individually did the same old comparison to the source via "align the samples exactly, inverting one waveform, mixing them together, and note the resulting value in `Amplify`:

32float: -94.494dB
24bit: -68.583dB
16bit: -24.057dB

And as a second sanity check of my own audio recording process, I even made a recording of the source 24bit FLAC in mpv with mpv's own internal volume set to 70% and did that same process of comparing & inverting waveforms, and I had a result of -45.551dB which isn't *amazing* but certainly more in-line with one would expect (also this is a nearly 4 year old version of mpv...)

　

All this is to say that Ren'Py still seems to be only handling around 16 bits of audio precision. This sort of thing is what I was concerned about with the inability to achieve bit-perfect audio even in a non-default configuration in that it would mask audio-rendering issues occurring somewhere in the pipeline just like it did for me initially.

**Comment by renpytom** [comment]
Okay, I'll reopen this for the bit-precision issue. 

I have no plans to try to make it bit-perfect, as I don't think that's something a game engine should try to do. It's not important compared to other goals. I would like to eventually decode to f32, but that's not a high priorty. (Right now, f32 is for mixing.)

**Comment by bober-gith** [comment]
For config.adjust_audio_amplitude not being the default, first, my understanding is that power panning exists so that when a sound is centered, because it's sent to both left and right speakers, it will lower the sides to about 0.7 so the combined loudness stays the same in the center instead of getting louder compared to the panned output. This means less manual work to balance audio if you use panning a lot.
But fully left or right sounds already stayed at 1.0 on one channel, so if you are trying boost everything back to full in the center using the config.adjust_audio_amplitude, as a side effect, in this new configuration, you might also distort and clip those already full panned sounds past their limit (rough example: center from 0.7 to 1.0, but full panned sounds from 1.0 to 1.3).
Imo this config shouldn't be used unless you know the risks (i'd avoid it entirely, current is better).
